units/event_log_tests.cc

```diff
@@ -0,0 +1,213 @@
+//    const int64_t create_time_ns = create_end - create_start;
+//    Int64Attr timestamp("timestamp", 1234);
+//    Int64Attr shader_hash("hash", res.shader_hash);
+//    std::vector<Attribute *> attributes = {&shader_hash, &timestamp};
+//    CreateShaderModuleEvent("create_shader_module_ns", attributes,
+//    LogLevel::High);
+// ```
+struct CreateShaderModuleEvent : Event {
+ public:
+  CreateShaderModuleEvent(const char *name, int64_t timestamp,
+                          int64_t hash_value, int64_t duration,
+                          LogLevel log_level)
+      : Event(name, log_level),
+        timestamp_{"timestamp", timestamp},
+        hash_value_{"hash", hash_value},
+        duration_{"duration", duration},
+        attributes_{&timestamp_, &hash_value_, &duration_} {}
+
+  const std::vector<Attribute *> &GetAttributes() override {
+    return attributes_;
+  }
+  size_t GetNumAttributes() const override { return attributes_.size(); }
+
+ private:
+  Int64Attr timestamp_;
+  Int64Attr hash_value_;
+  Int64Attr duration_;
+  std::vector<Attribute *> attributes_;
+};
+
+std::string ValueToCSVString(const std::string &value) { return value; }
+
+std::string ValueToCSVString(const int64_t value) {
+  return std::to_string(value);
+}
+
+std::string ValueToCSVString(const std::vector<int64_t> &values) {
+  std::stringstream csv_string;
+  csv_string << "\"[";
+  size_t e = values.size();
+  for (size_t i = 0; i != e; ++i) {
+    const char *delimiter = i < e - 1 ? "," : "";
+    csv_string << std::to_string(values[i]) + delimiter;
+  }
+  csv_string << "]\"";
+  return csv_string.str();
+}
+
+// Takes an `Event` instance as an input and generates a csv string containing
+// `event`'s name and attribute values.
+std::string EventToCSVString(Event &event) {
+  const std::vector<Attribute *> &attributes = event.GetAttributes();
+
+  std::stringstream csv_str;
+  csv_str << event.GetEventName();
+  csv_str << ",";
+  size_t e;
+  for (size_t i = 0, e = attributes.size(); i != e; ++i) {
+    switch (attributes[i]->GetValueType()) {
+      case ValueType::Int64: {
+        csv_str << ValueToCSVString(
+            attributes[i]->cast<Int64Attr>()->GetValue());
+        break;
+      }
+      case ValueType::String: {
+        csv_str << ValueToCSVString(
+            attributes[i]->cast<StringAttr>()->GetValue());
+        break;
+      }
+      case ValueType::VectorInt64: {
+        csv_str << ValueToCSVString(
+            attributes[i]->cast<VectorInt64Attr>()->GetValue());
+        break;
+      }
+    }
+    if (i + 1 != e) csv_str << ",";
+  }
+  return csv_str.str();
+}
+
+TEST(Event, AttributeCreation) {
+  const Int64Attr timestamp("timestamp", 1601314733266018318);
+  const StringAttr state("state", "1");
+  const VectorInt64Attr pipeline("pipeline",
+                                 {0x67d6fd0aaa78a6d8, 0x67d390249c2f20ce});
+  EXPECT_EQ(timestamp.GetName(), "timestamp");
+  EXPECT_EQ(timestamp.GetValue(), 1601314733266018318);
```

Comment:
It'd be better to put those long constants in local variables, so that it's immediately obvious when things should be equal, e.g.:
```c++
const int64_t timestamp_val = 160...;
const Int64Attr timestamp("timestamp", timestamp_val);
...
CHEKC_EQ(timestemp.GetValue(), timestamp_val);
```